src/platforms/atomic-kms/server/kms/quirks.cpp

```diff
@@ -56,100 +60,221 @@ class mga::Quirks::Impl
-
-        bool const should_skip_modesetting_support = skip_modesetting_support.count(driver);
-        if (should_skip_modesetting_support)
+        auto const driver = mgc::get_device_driver(parent_device.get());
+        mir::log_debug(
+            "Quirks(disable-kms-probe): checking device with devnode: %s, driver %s", device.devnode(), driver);
+
+        bool const should_skip_devnode = disable_kms_probe.skipped_devnodes.contains(devnode);
+        if (should_skip_devnode)
+        {
+            mir::log_info("Quirks(disable-kms-probe): skipping device %s (matches devnode quirk %s)", devnode, devnode);
+        }
+
+        bool const should_skip_driver = disable_kms_probe.skipped_drivers.contains(driver);
+        if (should_skip_driver)
         {
-            mir::log_info("Quirks: skipping modesetting check %s (matches driver quirk %s)", devnode, driver);
+            mir::log_info("Quirks(disable-kms-probe): skipping device %s (matches driver quirk %s)", devnode, driver);
         }
-        return !should_skip_modesetting_support;
+
+        return !(should_skip_driver || should_skip_devnode);
+    }
+
+    auto gbm_quirks_for(udev::Device const& device) -> std::shared_ptr<GbmQuirks>
+    {
+        class GbmCreateSurfaceNoFlags : public GbmQuirks::CreateSurfaceFlagsQuirk
+        {
+            auto gbm_create_surface_flags() const -> uint32_t override
+            {
+                return 0;
+            }
+        };
+
+        class GbmSurfaceHasFreeBuffersAlwaysTrue : public GbmQuirks::SurfaceHasFreeBuffersQuirk
+        {
+            auto gbm_surface_has_free_buffers(gbm_surface*) const -> int override
+            {
+                return 1;
+            }
+        };
+
+        class DefaultGbmCreateSurfaceFlags : public GbmQuirks::CreateSurfaceFlagsQuirk
+        {
+            auto gbm_create_surface_flags() const -> uint32_t override
+            {
+                return GBM_BO_USE_RENDERING | GBM_BO_USE_SCANOUT;
+            }
+        };
+
+        class DefaultGbmSurfaceHasFreeBuffers : public GbmQuirks::SurfaceHasFreeBuffersQuirk
+        {
+            auto gbm_surface_has_free_buffers(gbm_surface* gbm_surface) const -> int override
+            {
+                return ::gbm_surface_has_free_buffers(gbm_surface);
+            }
+        };
+
+        auto const driver = mgc::get_device_driver(device.parent().get());
+        auto const devnode = device.devnode();
+        mir::log_debug("Quirks(gbm-create-surface-flags + gbm-surface-has-free-buffers): checking device with devnode: %s, driver %s", devnode, driver);
+
+        auto const devnode_or_driver =
+            [](auto devnode, auto driver, auto devnodes, auto drivers) -> std::optional<std::string>
+        {
+            if (devnodes.contains(devnode))
+                return devnodes.at(devnode);
+
+            if (drivers.contains(driver))
+                return drivers.at(driver);
+
+            return std::nullopt;
+        };
```

Comment:
This template appears a couple of times and is (I think) only instantiated on one set of types. In particular `devnodes` and `drivers` are always `std::unordered_map<std::string, std::string>` and passed by value. Which is not efficient (I know this is not a hot path, but this is needless inefficiency).

Also, it isn't returning the result needed by the calling code: all invocation points use the result as a temporary and do the same post processing (apart from a string in the log message - which could be passed at the call site).

I feel that a refactor could both clean this logic up and silence the `bugprone-unchecked-optional-access` warning